### ISSUE TITLE
Numpy consistent aliases (clip/clamp, copy/clone)

### ISSUE THREAD

**Issue by avloss** [pytorch/pytorch#42053]
## 🚀 Feature
Add aliases for functions which are present in Numpy and perform similar (same?) functionality to PyTorch.

## Motivation

Too often one ends up jumping between PyTorch and Numpy (for plotting purposes, but also for scipy and other things). It's more than frequent that one encounters `AttributeError: 'Tensor' object has no attribute 'clip'`, when by inaccuracy applies function of the wrong name. It's understandable, that naming convention for PyTorch, perhaps, has a good motivation behind it, but adding convenience aliases could alleviate some of unnecessary pain.

## Pitch

Introducing aliases for some functionality found in NumPy. NumPy has 1+ version and a stable API, which is familiar to most Python users who work with numerical data.

## Alternatives

Adding a wrapper library, creating a library to monkey-patch the repo, or forking the repo. None of those are desirable, since the risk of breaking things would overweight any kind of convenience benefits. 

## Additional context

I understand that this is unlikely to be done, but simply wanted to see community opinion about this issue, and if, perhaps, I'm missing some understanding behind those differences in naming.

**Comment by kshitij12345** [comment]
Agreed. Many times one does naturally use the `numpy` API just for the code to crash.

AFAIK, @mruberry is looking over the aliasing of these equivalent functions.

**Comment by mruberry** [comment]
Great idea, @avloss. @kshitij12345 is correct, and we should deliver aliasing for the 1.7 release.

**Comment by rgommers** [comment]
It looks like there's no issue for aliases yet, so let's cross-link to the main tracking issue for NumPy compat (gh-38349) and leave this open.

+1 for aliases of sensible functions. Note that NumPy has aliases itself as well (e.g. `around`, `round`, `round_`), let's make sure to exclude those. It's sometimes unclear which one the main function is, it may not be the one that has the implementation attached to it. I can provide some usage statistics if needed that can help decide.

**Comment by mruberry** [comment]
> It looks like there's no issue for aliases yet, so let's cross-link to the main tracking issue for NumPy compat ([gh-38349](https://github.com/pytorch/pytorch/issues/38349)) and leave this open.
> 
> +1 for aliases of sensible functions. Note that NumPy has aliases itself as well (e.g. `around`, `round`, `round_`), let's make sure to exclude those. It's sometimes unclear which one the main function is, it may not be the one that has the implementation attached to it. I can provide some usage statistics if needed that can help decide.

Usage stats would be very helpful if we're going to be selective about which aliases we implement.

The only alias we have today is "absolute" for "abs," which was implemented as a prototype so we could discuss how we wanted to implement aliases in the future. It's not hard to write them but I think we might want to simplify the process even further.

**Comment by zou3519** [comment]
If there is any debate or design decision over how aliases should work, here is a data point and an opinion. From a Tensor extension point of view, it is nicer for aliases to not be defined as operators. 

A tensor extension generally has to override all operators (or most operators). Adding aliases into native functions (like how absolute is implemented) increases the surface area for tensor extensions. It would be nice if aliases weren't overridable operators and instead translate directly to a call to an operator.
https://github.com/pytorch/pytorch/blob/4cdbe5c4952ee6b7be24ff15b1d45d0fb593b13f/aten/src/ATen/native/native_functions.yaml#L227-L240

**Comment by mruberry** [comment]
> A tensor extension generally has to override all operators (or most operators). Adding aliases into native functions (like how absolute is implemented) increases the overridable API surface area for tensor extensions. It would be nice if aliases weren't overridable operators and instead translate directly to a call to an operator.

That's an excellent point but I don't see how we can avoid it completely. Just focusing on Python for the moment, we could naturally define aliases like this:

```
foo_alias = foo
```

But if an extension monkey patches `foo` they'll still have to update `foo_alias`. If they override the actual mapping from Python bindings to C++ functions then they'd be OK, though.

**Comment by zou3519** [comment]
For the Python example, I would argue that monkey-patching is different from dispatching to different implementations. A tensor extension wouldn't monkey-patch `foo`; `foo` should contain some logic that dispatches to the implementation requested by the tensor extension. I agree there's no way to support aliasing with monkey-patching, as your code example shows.

In C++, using `absolute` as an example, could we hand-write (or codegen) `at::absolute`, `Tensor::absolute` functions that call into dispatchable operators? `at::absolute` would itself not be a dispatchable operator, but `at::abs` would be. e.g.,

```
namespace at {

Tensor absolute(const Tensor& self) {
   // at::abs does the dynamic dispatch. If someone overrides abs for a tensor extension, then at::absolute still works
   return at::abs(self);
}
Tensor& absolute(Tensor& out, const Tensor& self) {
   // at::abs does the dynamic dispatch. If someone overrides abs for a tensor extension, then absolute still works
   return at::abs(out, self);
}
// Etc.

}
```

**Comment by mruberry** [comment]
I was also thinking that, although maybe we can just do:

```
const auto absolute = at::abs;
```

analogous to Python.

**Comment by mruberry** [comment]
See https://github.com/pytorch/pytorch/pull/42586. That PR actually implements aliases using redispatch as you suggested, @zou3519, because I thought implementing it using language features would result in a proliferation of lists. The PR enumerates some pros/cons of both approaches.

**Comment by mruberry** [comment]
https://github.com/pytorch/pytorch/pull/42770 for the clip alias.

Copy is interesting. We already have [torch.Tensor.copy_](https://pytorch.org/docs/master/tensors.html?highlight=copy#torch.Tensor.copy_), and it and clone have different autograd semantics. NumPy doesn't have autograd semantics, of course. What would you expect when calling `torch.copy(t)`, would the copy work like `t.clone()` and appear in the autograd graph?

**Comment by mruberry** [comment]
Updating title of this issue because we have clip: https://pytorch.org/docs/master/generated/torch.clip.html?highlight=clip#torch.clip.
